src/s3ql/backends/backblaze.py

```diff
@@ -0,0 +1,704 @@
+
+        j = json.loads(str(body, encoding='UTF-8'))
+        if j['files'][0]['fileName'] == key_with_prefix:
+            return j['files'][0]['contentLength']
+
+        raise NoSuchObject(key_with_prefix)
+
+    @retry
+    @copy_ancestor_docstring
+    def open_read(self, key):
+        log.debug('started with %s', key)
+
+        key = self._encode_key(key)
+
+        resp, data = self._do_request('GET',
+                                      '/file/%s/%s%s' % (self.bucket_name, self.prefix, key),
+                                      self.conn_download, download_body=False)
+
+        meta = self._extractmeta(resp, key)
+
+        return ObjectR(key, self.conn_download, resp, metadata=meta)
+
+    @copy_ancestor_docstring
+    def open_write(self, key, metadata=None, is_compressed=False):
+        log.debug('started with %s', key)
+
+        key = self._encode_key(key)
+
+        return ObjectW(key, self, metadata)
+
+    @retry
+    @copy_ancestor_docstring
+    def clear(self):
+        log.debug('started')
+        for file in self.list():
+            self.delete(file,force=True)
+
+    @retry
+    @copy_ancestor_docstring
+    def delete(self, key, force=False):
+        log.debug('started with %s', key)
+
+        todel_id = self._list_file_version(key)
+
+        if not todel_id and not force:
+            raise NoSuchObject(key)
+
+        self._delete_file_id(todel_id)
+
+    @copy_ancestor_docstring
+    def list(self, prefix=''):
+        log.debug('started with %s', prefix)
+
+        prefix = self._encode_key(prefix)
+
+        next_filename = self.prefix + prefix
+        keys_remaining = True
+
+        while keys_remaining and next_filename is not None:
+            next_filename, filelist = self._list_file_name(next_filename)
+
+            while filelist:
+                file = filelist.pop(0)
+
+                if file.startswith(self.prefix + prefix):
+                    # remove prefix before return
+                    r = file[len(self.prefix):]
+                    yield self._decode_key(r)
+                else:
+                    keys_remaining = False
+                    break
+
+    @prepend_ancestor_docstring
+    def copy(self, src, dest, metadata=None):
+        """No atomic copy operation on backend. Hope this does not append to often"""
+        log.debug('started with %s, %s', src, dest)
+
+        data, src_meta = self.fetch(src)
+
+        # Delete dest file if already exist
+        self.delete(dest, force=True)
+
+        if metadata is None:
+            dst_meta = src_meta
+        else:
+            dst_meta = metadata
+
+        self.store(dest, data, dst_meta)
+
+    @copy_ancestor_docstring
+    def update_meta(self, key, metadata):
+        log.debug('started with %s', key)
+        self.copy(key, key, metadata)
+
+    def close(self):
+        log.debug('started')
+        self.conn_api.disconnect()
+        self.conn_download.disconnect()
+
+
+class ObjectR(object):
```

Comment:
There seems to be a lot of code duplication from s3.ObjectR here. Can you just inherit from s3.ObjectR and change the base class so that the hash function and response header name can be easily customized?